vm/src/vm/vm_core.rs

```diff
@@ -493,7 +499,14 @@ impl VirtualMachine {
         let ap = self.run_context.get_ap();
         let output_address = self.segments.memory.get_relocatable(ap)?;
 
-        let new_state = blake2s_compress(&state, &message, counter, 0, 0, 0);
+        let new_state = blake2s_compress(
+            &state,
+            &message,
+            counter,
+            0,
+            if !is_last_block { 0 } else { 0xffffffff },
```

Comment:
Please extract this logic into a variable next to the definitions of `state` and `message` (lines 481-497)